modules/async_message.py

```diff
@@ -0,0 +1,58 @@
+"""
+This module contains classes for aiding in processing incoming and outgoing asyncronous messages
+"""
+import threading
+import queue
+import time
+
+class AsyncMessageCallback:
+    """
+    This class provides an asyncronous interface for recieving and queue messages
+    """
+    def __init__(self, wait_timeout=None, queue_size=-1):
+        self._wait_timeout = wait_timeout
+
+        self._message_queue = queue.Queue(maxsize=queue_size)
+        self._new_message_event = threading.Event()
+
+    def wait_for_message(self):
+        """
+        Blocking message retrieval
+        """
+        self._new_message_event.wait(timeout=self._wait_timeout)
+        msg = None
+
+        try:
+            msg = self._message_queue.get_nowait()
+        except queue.Empty:
+            self._clear_message_event()
+
+        return msg
+
+    def queue_message(self, data):
+        """
+        Add an incoming message to the queue and flag
+        """
+        try:
+            self._message_queue.put_nowait((time.time(), data))
+            self._set_message_event()
+        except queue.Full:
+            pass # Silently drop message
```

Comment:
Probably want to log at least some exception here for if the queue is full. Silent failures are very bad juju

Reply:
That's a fair point, will probably be necessary to see if lots of messages are being dropped and reduce the production rate

Reply:
I've added in some TODO notes that this needs to be fixed in the future